internal/configtxgen/genesisconfig/config.go

```diff
@@ -217,21 +231,48 @@ func LoadTopLevel(configPaths ...string) *TopLevel {
 
 	err := config.ReadInConfig()
 	if err != nil {
-		logger.Panic("Error reading configuration: ", err)
+		logger.Panicf("Error reading configuration: %s", err)
 	}
 	logger.Debugf("Using config file: %s", config.ConfigFileUsed())
 
-	var uconf TopLevel
-	err = viperutil.EnhancedExactUnmarshal(config, &uconf)
+	uconf, err := cfgCache.loadConfigCache(config, config.ConfigFileUsed())
 	if err != nil {
-		logger.Panic("Error unmarshaling config into struct: ", err)
+		logger.Panicf("failed to load configCache: %s", err)
 	}
+	logger.Infof("Loaded configuration: %s", config.ConfigFileUsed())
 
-	(&uconf).completeInitialization(filepath.Dir(config.ConfigFileUsed()))
+	return uconf
+}
 
-	logger.Infof("Loaded configuration: %s", config.ConfigFileUsed())
+// loadConfigCache loads the TopLevel config structure from configCache.
+// if not successful, it unmarshal a config file, and populate configCache
+// with marshaled TopLevel struct.
+func (cache *configCache) loadConfigCache(config *viper.Viper, configPath string) (*TopLevel, error) {
```

Comment:
nit: just rename the self-referential "cache" to "c" so we see `c.cache` instead of `cache.cache`.

please do this separately

Reply:
done